executor/src/witgen/machines/write_once_memory.rs

```diff
@@ -0,0 +1,221 @@
+use std::collections::{BTreeMap, HashMap};
+
+use ast::{
+    analyzed::{
+        AlgebraicExpression as Expression, AlgebraicReference, Identity, IdentityKind, PolyID,
+        PolynomialType,
+    },
+    parsed::SelectedExpressions,
+};
+use number::{DegreeType, FieldElement};
+
+use crate::witgen::{
+    affine_expression::AffineExpression, util::try_to_simple_poly, EvalResult, EvalValue,
+    FixedData, IncompleteCause, MutableState, QueryCallback,
+};
+
+use super::{FixedLookup, Machine};
+
+/// A machine that has a fixed address space and can be written to once.
```

Comment:
```suggestion
/// A machine that has a fixed address space, each address can be written to once.
```
Hm, this is also not too accurate, since we can write multiple times, it just has to be the same value.
But maybe we shouldn't be too pedantic here.